nextcord/http.py

```diff
@@ -240,151 +707,280 @@ async def request(
+            route.method,
+            route.bucket,
+            _get_logging_auth(auth),
+        )  # Only use this for logging.
+        ret: Union[Optional[str], dict] = None
         response: Optional[aiohttp.ClientResponse] = None
-        data: Optional[Union[Dict[str, Any], str]] = None
-        await lock.acquire()
-        with MaybeUnlock(lock) as maybe_lock:
-            for tries in range(5):
-                if files:
-                    for f in files:
-                        f.reset(seek=tries)
-
-                if form:
-                    form_data = aiohttp.FormData(quote_fields=False)
-                    for params in form:
-                        form_data.add_field(**params)
-                    kwargs["data"] = form_data
-
-                try:
-                    async with self.__session.request(method, url, **kwargs) as response:
+
+        # If retry_request is False and any of the rate limits are locked, don't continue and raise immediately.
+        if retry_request is False:
+            if global_rate_limit.locked:
+                _log.info(
+                    "Path %s was called with retry_request=False while the global rate limit is locked.",
+                    rate_limit_path,
+                )
+                raise HTTPInternalRatelimitLocked(
+                    "Request would exceed the global ratelimit and retries are disabled."
+                )
+
+            if url_rate_limit.locked:
+                _log.info(
+                    "Path %s was called with retry_request=False while the URL rate limit is locked.",
+                    rate_limit_path,
+                )
+                raise HTTPInternalRatelimitLocked(
+                    "Request would exceed the route ratelimit and retries are disabled."
+                )
+
+        # The loop is to allow migration to a different RateLimit if needed.
+        # If we hit this loop max_retry_count times, something is wrong. Either we're migrating buckets way
+        #  too much, 429s keep getting hit, or something is internally wrong.
+        for retry_count in range(max_retry_count):  # To prevent infinite loops.
+            should_retry = False
+            try:
+                async with global_rate_limit, url_rate_limit:
+                    # This check is for asyncio.gather()'d requests where the rate limit can change.
+                    if (
+                        temp := self._get_url_rate_limit(route.method, route, auth)
+                    ) is not url_rate_limit and temp is not None:
+                        _log.debug(
+                            "Route %s had the rate limit changed, resetting and retrying.",
+                            rate_limit_path,
+                        )
+                        url_rate_limit = temp
+                        continue
+
+                    if files:
+                        for f in files:
+                            f.reset(seek=retry_count)
```

Comment:
Wouldn't make this so files are offset by one byte for every retry?

Reply:
idk, I just took it from the original HTTPClient code.

Reply:
```suggestion
                            f.reset(seek=bool(retry_count))
```

Who came up with this 😭

The value of seek isn't interesting, it's whether it's truthy (≠0) or falsy (=0).
→ If we are retrying/retried, we seek back to the original position.